src/com/amazon/corretto/crypto/provider/Loader.java

```diff
@@ -177,63 +162,112 @@ static String getProperty(String propertyName, String def) {
+         *
+         * Any shared libraries that are not present in Java's LOCAL object group (Group #1), will be loaded into a
+         * new child LOCAL object group, lower in the object group hierarchy. This can be done multiple times for
+         * multiple different JNI libraries. Note that since both Groups #2 and #3 have Group #1 as a parent, their
+         * library symbols will not conflict with each another since each group only see's symbols from their own
+         * parents in the hierarchy above them. This means it is possible for multiple different JNI libraries to be
+         * loaded at runtime that use different libcrypto implementations so long as those JNI libraries configure
+         * their RPath values correctly, and are compatible with any libraries that have already been loaded above
+         * them in the hierarchy.
+         *
+         * Once a LOCAL object group is created and the recursive library loading is complete, that LOCAL object group
+         * can no longer be modified at runtime other than to be deleted entirely with dlclose(). Subsequent
+         * System.load() or dlopen(..., RTLD_LOCAL) calls will only create new child LOCAL Object Groups below the
+         * current object group in the hierarchy, and will only load in shared libraries that are not present in the
+         * caller's object group's hierarchy.
+         *
+         * Links:
+         *  - https://docs.oracle.com/cd/E19957-01/806-0641/6j9vuquj2/index.html
+         *  - http://people.redhat.com/drepper/dsohowto.pdf
+         *  - https://docs.oracle.com/cd/E23824_01/pdf/819-0690.pdf
+         *  - https://man7.org/linux/man-pages/man3/dlopen.3.html
+         */
+        Path accpJniSharedLibraryPath = privateTempDirectory.resolve(System.mapLibraryName(JNI_LIBRARY_NAME)).toAbsolutePath();
+        System.load(accpJniSharedLibraryPath.toString());
+
+        // If loading library from JAR file, then the compile-time and run-time libcrypto versions should be an exact match.
+        validateLibcryptoExactVersionMatch();
+
+        maybeDeletePrivateTempDir(privateTempDirectory);
+    }
 
-        final Path libPath = createTmpFile(prefix, suffix);
 
-        try (InputStream is = Loader.class.getResourceAsStream(resourceName);
-                OutputStream os = Files.newOutputStream(libPath, StandardOpenOption.CREATE,
-                        StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)) {
-            final byte[] buffer = new byte[16 * 1024];
-            int read = is.read(buffer);
-            while (read >= 0) {
-                os.write(buffer, 0, read);
-                read = is.read(buffer);
-            }
-            os.flush();
-        }
-        return libPath;
+    private static void tryLoadLibraryFromSystem() {
+        /**
+         * Attempt to load library using system's default shared library lookup paths
```

Comment:
Should we still call `validateLibcryptoExactVersionMatch();`? After all, if the versions don't match, how do you know we'll work and not crash horribly?

Do we need a fuzzier match for the system library case?

Reply:
I'm assuming that if someone is manually loading ACCP from system libraries, then the person installing the libraries into the system directories will be ensuring that ACCP and LibCrypto work together, so I added the fuzzier version check if loading ACCP from system libraries. 

In practice however, AWS-LC is not regularly updating `OPENSSL_VERSION_NUMBER` so this check may not provide as much value as I initially intended. I've opened a PR to AWS-LC to provide an API to retrieve their `AWSLC_API_VERSION` value at runtime